bootloader/piboot.go

```diff
@@ -0,0 +1,416 @@
+}
+
+func (p *piboot) processBlOpts(blOpts *Options) {
+	if blOpts == nil {
+		return
+	}
+
+	switch {
+	case blOpts.Role == RoleRecovery || blOpts.NoSlashBoot:
+		if !blOpts.PrepareImageTime {
+			p.rootdir = ubuntuSeedDir
+		}
+		// RoleRecovery or NoSlashBoot imply we use
+		// the environment file in /piboot/ubuntu as
+		// it exists on the partition directly
+		p.basedir = pibootPartFolder
+	}
+}
+
+// newPiboot creates a new Piboot bootloader object
+func newPiboot(rootdir string, blOpts *Options) Bootloader {
+	p := &piboot{
+		rootdir: rootdir,
+	}
+	p.setDefaults()
+	p.processBlOpts(blOpts)
+	return p
+}
+
+func (p *piboot) Name() string {
+	return "piboot"
+}
+
+func (p *piboot) dir() string {
+	if p.rootdir == "" {
+		panic("internal error: unset rootdir")
+	}
+	return filepath.Join(p.rootdir, p.basedir)
+}
+
+func (p *piboot) envFile() string {
+	return filepath.Join(p.dir(), pibootCfgFilename)
+}
+
+// piboot enabled if env file exists
+func (p *piboot) Present() (bool, error) {
+	return osutil.FileExists(p.envFile()), nil
+}
+
+// Variables stored in ubuntu-seed:
+//   snapd_recovery_system
+//   snapd_recovery_mode
+//   snapd_recovery_kernel
+// Variables stored in ubuntu-boot:
+//   kernel_status
+//   snap_kernel
+//   snap_try_kernel
+//   snapd_extra_cmdline_args
+//   snapd_full_cmdline_args
+//   recovery_system_status
+//   try_recovery_system
+func (p *piboot) SetBootVars(values map[string]string) error {
+	env, err := ubootenv.OpenWithFlags(p.envFile(), ubootenv.OpenBestEffort)
+	if err != nil {
+		return err
+	}
+
+	// Set when we change a boot env variable, to know if we need to save the env
+	dirtyEnv := false
+	// Flag to know if we need to write piboot's config.txt or tryboot.txt
+	reconfigBootloader := false
+	for k, v := range values {
+		// already set to the right value, nothing to do
+		if env.Get(k) == v {
+			continue
+		}
+		env.Set(k, v)
+		dirtyEnv = true
+		// Cases that change the bootloader configuration
+		if k == "snapd_recovery_mode" || k == "kernel_status" {
+			reconfigBootloader = true
+		}
+		if k == "snap_try_kernel" && v == "" {
+			// refresh (ok or not) finished, remove tryboot.txt
+			// (os_prefix in config.txt will be changed now in
+			// loadAndApplyConfig in the ok case)
+			trybootPath := filepath.Join(ubuntuSeedDir, "tryboot.txt")
+			if err := os.Remove(trybootPath); err != nil {
+				logger.Noticef("cannot remove %s: %v", trybootPath, err)
+			}
+		}
+	}
+
+	if dirtyEnv {
+		if err := env.Save(); err != nil {
+			return err
+		}
+	}
+
+	if reconfigBootloader {
+		if err := p.loadAndApplyConfig(env); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *piboot) SetBootVarsFromInitramfs(values map[string]string) error {
+	env, err := ubootenv.OpenWithFlags(p.envFile(), ubootenv.OpenBestEffort)
+	if err != nil {
+		return err
+	}
+
+	dirtyEnv := false
+	for k, v := range values {
+		// already set to the right value, nothing to do
+		logger.Noticef("Setting %s=%s", k, v)
```

Comment:
do we need this logging?

Reply:
Good catch, it was some debugging left-over. Removed now.

Reply:
this still seems to be here?